src/ray/raylet/scheduling/policy/scheduling_policy.h

```diff
@@ -17,24 +17,86 @@
 
 namespace ray {
 namespace raylet_scheduling_policy {
 
+// Status of resource scheduling result.
+struct SchedulingResultStatus {
+  bool IsFailed() const { return code == SchedulingResultStatusCode::FAILED; }
+  bool IsInfeasible() const { return code == SchedulingResultStatusCode::INFEASIBLE; }
+  bool IsSuccess() const { return code == SchedulingResultStatusCode::SUCCESS; }
+
+  enum class SchedulingResultStatusCode {
+    // Scheduling failed but retryable.
+    FAILED = 0,
+    // Scheduling failed and non-retryable.
+    INFEASIBLE = 1,
+    // Scheduling successful.
+    SUCCESS = 2,
+  };
+  SchedulingResultStatusCode code = SchedulingResultStatusCode::SUCCESS;
+};
+
+struct SchedulingResult {
+  static SchedulingResult Infeasible() {
+    SchedulingResult result;
+    result.status.code = SchedulingResultStatus::SchedulingResultStatusCode::INFEASIBLE;
+    return result;
+  }
+
+  static SchedulingResult Failed() {
+    SchedulingResult result;
+    result.status.code = SchedulingResultStatus::SchedulingResultStatusCode::FAILED;
+    return result;
+  }
+
+  static SchedulingResult Success(std::vector<scheduling::NodeID> &&nodes) {
+    SchedulingResult result;
+    result.status.code = SchedulingResultStatus::SchedulingResultStatusCode::SUCCESS;
+    result.selected_nodes = std::move(nodes);
+    return result;
+  }
+
+  // The status of scheduling.
+  SchedulingResultStatus status;
+  // The nodes successfully scheduled.
+  std::vector<scheduling::NodeID> selected_nodes;
+};
+
 /// ISchedulingPolicy picks a node to from the cluster, according to the resource
 /// requirment as well as the scheduling options.
 class ISchedulingPolicy {
  public:
   virtual ~ISchedulingPolicy() = default;
   /// \param resource_request: The resource request we're attempting to schedule.
-  /// \param scheduling_options: scheduling options.
+  /// \param options: scheduling options.
   ///
   /// \return NodeID::Nil() if the task is unfeasible, otherwise the node id
   /// to schedule on.
   virtual scheduling::NodeID Schedule(const ResourceRequest &resource_request,
-                                      SchedulingOptions options) = 0;
+                                      SchedulingOptions options) {
+    return scheduling::NodeID();
+  }
+
+  /// Schedule the specified resources to the cluster nodes.
+  ///
+  /// \param resource_request_list The resource request list we're attempting to schedule.
+  /// \param options: scheduling options.
+  /// \param context: The context of current scheduling. Each policy can
+  /// correspond to a different type of context.
+  /// \return `SchedulingResult`, including the
+  /// selected nodes if schedule successful, otherwise, it will return an empty vector and
+  /// a flag to indicate whether this request can be retry or not.
+  virtual SchedulingResult Schedule(
+      const std::vector<const ResourceRequest *> &resource_request_list,
+      SchedulingOptions options,
+      SchedulingContext *context) {
```

Comment:
will a reference be better here?

Reply:
> will a reference be better here?

It can't be `const SchedulingContext &` here if the `selected_nodes` is added to the `context`, because it is mutable.